src/script/services/publish/android-publish.ts

```diff
@@ -0,0 +1,177 @@
+import { getManifest, getManiURL } from '../manifest';
+
+export async function generateAndroidPackage(
+  androidOptions: AndroidApkOptions
+) {
+  const validationErrors = validateAndroidOptions(androidOptions);
+  if (validationErrors.length > 0 || !androidOptions) {
+    throw new Error(
+      'Invalid Adroid options. ' + validationErrors.map(a => a.error).join('\n')
+    );
+  }
+
+  const generateApkUrl = `${env.androidPackageGeneratorUrl}/generateApkZip`;
+
+  try {
+    const response = await fetch(generateApkUrl, {
+      method: 'POST',
+      headers: new Headers({ 'content-type': 'application/json' }),
+      body: JSON.stringify(androidOptions),
+    });
+    if (response.status === 200) {
+      const data = await response.blob();
+
+      await fileSave(data, {
+        fileName: 'your_android_pwa.zip',
+        extensions: ['.zip'],
+      });
+    } else {
+      const responseText = await response.text();
+
+      // Did it fail because images couldn't be fetched with ECONNREFUSED? E.g. https://github.com/pwa-builder/PWABuilder/issues/1312
+      // If so, retry using our downloader proxy service.
+      const hasSafeImages =
+        androidOptions.iconUrl &&
+        androidOptions.iconUrl.includes(env.safeUrlFetcher || '');
+      const isConnectionRefusedOrForbidden =
+        (responseText || '').includes('ECONNREFUSED') ||
+        response.status === 403;
+
+      if (!hasSafeImages && isConnectionRefusedOrForbidden) {
+        console.warn(
+          'Android package generation failed with ECONNREFUSED. Retrying with safe images.',
+          responseText
+        );
+        this.updateAndroidOptionsWithSafeUrls(androidOptions);
+
+        await generateAndroidPackage(androidOptions);
+      } else {
+        throw new Error(
+          `Error generating Android package.\n\nStatus code: ${response.status}\n\nError: ${response.statusText}\n\nDetails: ${responseText}`
+        );
+      }
+    }
+  } catch (err) {
+    throw new Error(
+      `Error generating Android platform due to HTTP error.\n\nStatus code: ${err.status}\n\nError: ${err.statusText}\n\nDetails: ${err}`
+    );
+  }
+}
+
+export function createAndroidPackageOptionsFromManifest(): AndroidApkOptions {
+  const manifest = getManifest();
+
+  if (manifest) {
+    const pwaUrl = getManiURL();
+
+    if (!pwaUrl) {
+      throw new Error("Can't find the current URL");
+    }
+
+    const appName = manifest.short_name || manifest.name || 'My PWA';
+    const packageName = generatePackageId(new URL(pwaUrl).hostname);
+    // Use standalone display mode unless the manifest has fullscreen specified.
+    const display =
+      manifest.display === 'fullscreen' ? 'fullscreen' : 'standalone';
+    // StartUrl must be relative to the host.
+    // We make sure it is below.
+    let relativeStartUrl: string;
+    if (
+      !manifest.start_url ||
+      manifest.start_url === '/' ||
+      manifest.start_url === '.' ||
+      manifest.start_url === './'
```

Comment:
would make for a nice validity/assert function, easy to reuse in tests and also elsewhere!